doc/decisions/reference_counting.md

```diff
@@ -0,0 +1,39 @@
+# Reference Counting
+
+## Problem
+
+- locking is not reset when ref counting again gets 0 (adding to keyset and
+  pop again) #2202
+- C++ API for KeySet and Key has unexpected differences: also use ref counting
+  for KeySets (also suggested in #1332)
+
+## Constraints
+
+## Assumptions
+
+## Considered Alternatives
+
+- make ref counting thread safe (probably useful for JNI)
+- start with 1 for reference counting and let keyDecRef do keyDel
+
+## Decision
+
+- add second reference counter to Key
+- One counter is for locking references, the other one for general references. A locking reference automatically locks/unlocks the keyname.
+- introduce reference counter for KeySets (for external keyset references)
+- `keyIncRef` (or equivalent) return error in case of reference counter overflow
```

Comment:
Let us settle for keyBorrow (see #3949) unless someone comes up with a better name (maybe keyShare). The rename should not be too costly, keyIncRef is not heavily used.